lecture-notes/2-vcs.md

```diff
@@ -420,11 +475,607 @@ commits; the general write-add-commit flow; that all Git objects are stored in
+   glance.
+1. The body is line-wrapped at 72 characters and organized into paragraphs.
+   References to other commits in the body take the form `commit 123456789abc
+   ("Fix foobar")`, where `123456789abc` is the first 12 characters of the other
+   commit's hash and "Fix foobar" is its title.
+
+Not everyone likes these particular conventions, and many projects set their own
+commit message style. Nonetheness, most projects agree that consistent styling
+across commits is a good thing, as it makes commit histories easier to read and
+makes collaboration easier.
+
+### Staging files partially (`git add -p`)
+If you try to follow our advice using the commands we've covered so far, you'll
+soon find that it's very difficult to create tidy commits when you can only
+stage changes on a file-by-file basis (`git add <file>`). The process of
+software development is rarely as organized as a Git history should be: while
+working on one change, for example, you might see other issues in the same file
+and fix them on the fly.
+
+But doesn't that force you to put both your original change and your quick fix
+in the same commit, since `git add` stages every change in the file? You might
+already be thinking of clever workarounds---like making a temporary copy of the
+file and then undoing one of the changes---but there's a better solution.
+
+`git add -p` is that solution, but Git beginners often don't even know it
+exists. Unlike plain `git add` (or worse, `git commit -a`, which commits every
+change in every file), `git add -p` lets you partially stage a file. When
+invoked, it displays each block of modified lines (called a *hunk*) in sequence
+and asks you whether or not to stage it[^add-p-editing]. As usual, a subsequent
+`git commit` will include only the staged hunks, letting you quickly and easily
+distribute your changes across multiple commits.
+
+[^add-p-editing]: You can also choose to edit a hunk before staging it, which
+    will cause `git add -p` to open a text editor where you can modify or delete
+    lines. This isn't typically necessary, but it helps when unrelated changes
+    are close enough together that Git counts them as a single hunk. (For
+    changes that are close but not directly adjacent, you can also ask `git add
+    -p` split the hunk into multiple smaller hunks and prompt for each.)
+
+```
+TODO: Example of `git add -p` to stage specific hunks that have changed.
+Demonstrate committing multiple changes separately with good messages.
+```
+
+### Bringing it all together
+
+With `git add -p` and some commit guidelines under your belt, you have
+everything you need to start tracking a project's history in Git. Here's how
+that might look for a simple project (which will look familiar once you begin
+Homework 4!). First, you create a repository:
+
+```console
+$ mkdir calc
+$ cd calc
+$ git init
+Initialized empty Git repository in /home/you/calc/.git/
+```
+
+Then, you commit some initial code:
```

Comment:
before each code blob please expressly call out the format and what keywords to look for so people have at least a vague idea of how to mentally parse